mbedtls/src/ssl/context.rs

```diff
@@ -500,6 +511,250 @@ impl HandshakeContext {
+where
+    AsyncIoAdapter<T>: IoCallbackUnsafe<AsyncStream>,
+{
+    fn poll_write(
+        mut self: Pin<&mut Self>,
+        cx: &mut TaskContext<'_>,
+        buf: &[u8],
+    ) -> Poll<IoResult<usize>> {
+        if self.handle().session.is_null() {
+            return Poll::Ready(Err(IoError::new(
+                IoErrorKind::Other,
+                "stream has been shutdown",
+            )));
+        }
+
+        let buf = {
+            let io = self
+                .io_mut()
+                .ok_or(IoError::new(IoErrorKind::Other, "stream has been shutdown"))?;
+            io.ecx.set(cx);
+            io.write_tracker.adjust_buf(buf)
+        }?;
+
+        self.io_mut()
+            .ok_or(IoError::new(IoErrorKind::Other, "stream has been shutdown"))?
+            .ecx
+            .set(cx);
+
+        let result = match unsafe {
+            ssl_write((&mut *self).into(), buf.as_ptr(), buf.len()).into_result()
+        } {
+            Err(Error::SslPeerCloseNotify) => Poll::Ready(Ok(0)),
+            Err(Error::SslWantWrite) => Poll::Pending,
+            Err(e) => Poll::Ready(Err(crate::private::error_to_io_error(e))),
+            Ok(i) => Poll::Ready(Ok(i as usize)),
+        };
+
+        let io = self
+            .io_mut()
+            .ok_or(IoError::new(IoErrorKind::Other, "stream has been shutdown"))?;
+
+        io.ecx.clear();
+        io.write_tracker.post_write(buf, &result);
+
+        cx.waker().clone().wake();
+
+        result
+    }
+
+    fn poll_flush(mut self: Pin<&mut Self>, cx: &mut TaskContext<'_>) -> Poll<IoResult<()>> {
+        // We can only flush the actual IO here.
+        // To flush mbedtls we need writes with the same buffer until complete.
+        let io = &mut self
+            .io_mut()
+            .ok_or(IoError::new(IoErrorKind::Other, "stream has been shutdown"))?
+            .inner;
+        let stream = Pin::new(io);
+        stream.poll_flush(cx)
+    }
+
+    fn poll_shutdown(mut self: Pin<&mut Self>, cx: &mut TaskContext<'_>) -> Poll<IoResult<()>> {
+        if self.handle().session.is_null() {
+            return Poll::Ready(Err(IoError::new(
+                IoErrorKind::Other,
+                "stream has been shutdown",
+            )));
+        }
+
+        self.io_mut()
+            .ok_or(IoError::new(IoErrorKind::Other, "stream has been shutdown"))?
+            .ecx
+            .set(cx);
+
+        let result = match unsafe { ssl_close_notify((&mut *self).into()).into_result() } {
+            Err(Error::SslWantRead) | Err(Error::SslWantWrite) => Poll::Pending,
+            Err(e) => {
+                unsafe {
+                    ssl_set_bio(
+                        (&mut *self).into(),
+                        ::core::ptr::null_mut(),
+                        None,
+                        None,
+                        None,
+                    );
+                }
+                self.io = None;
```

Comment:
This code is repeated 3×, please find a way to consolidate.

Reply:
refactored